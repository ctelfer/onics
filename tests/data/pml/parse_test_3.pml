# Constant and inline optimization test
#
const A = 10;
const B = 5;
const C = A + B;
const D = A - B;
const E = A * B;
const F = A / B;
const G = A % B;


inline square(x) { x * x }

const H = square(B);
const I = square(square(B));

inline wierd(i, j) { i + -j }

const J = wierd(E, A);
const K = wierd(square(A), wierd(E, A));

const L = pop(0x1818181818181818);
const M = min(300, -5);
const N = max(300, -5);
const O = log2(0x4321);
