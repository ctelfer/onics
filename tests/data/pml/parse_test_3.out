#########
Initializing parse data structures
#########
#########
Starting Parse
#########


########
Done parsing, destroying scanner and parser
########


########
Optimizing tree:
########
Done... printing optimized tree
Printing PML Abstract Syntax Tree
-----------
Variables
-----------
   Variable: mem [global; width=0, addr=0]
   Variable: A [const; width=8, addr=0]
      Initialization value -- 
      Scalar [[c]; width=8] -- width 8: 10 (10,0xa)
   Variable: B [const; width=8, addr=0]
      Initialization value -- 
      Scalar [[c]; width=8] -- width 8: 5 (5,0x5)
   Variable: C [const; width=8, addr=0]
      Initialization value -- 
      Scalar [[c,p]; width=8] -- width 8: 15 (15,0xf)
   Variable: D [const; width=8, addr=0]
      Initialization value -- 
      Scalar [[c,p]; width=8] -- width 8: 5 (5,0x5)
   Variable: E [const; width=8, addr=0]
      Initialization value -- 
      Scalar [[c,p]; width=8] -- width 8: 50 (50,0x32)
   Variable: F [const; width=8, addr=0]
      Initialization value -- 
      Scalar [[c,p]; width=8] -- width 8: 2 (2,0x2)
   Variable: G [const; width=8, addr=0]
      Initialization value -- 
      Scalar [[c,p]; width=8] -- width 8: 0 (0,0x0)
   Variable: H [const; width=8, addr=0]
      Initialization value -- 
      Scalar [[c,p]; width=8] -- width 8: 25 (25,0x19)
   Variable: I [const; width=8, addr=0]
      Initialization value -- 
      Scalar [[c,p]; width=8] -- width 8: 625 (625,0x271)
   Variable: J [const; width=8, addr=0]
      Initialization value -- 
      Scalar [[c,p]; width=8] -- width 8: 40 (40,0x28)
   Variable: K [const; width=8, addr=0]
      Initialization value -- 
      Scalar [[c,p]; width=8] -- width 8: 60 (60,0x3c)
   Variable: L [const; width=8, addr=0]
      Initialization value -- 
      Scalar [[c,p]; width=8] -- width 8: 16 (16,0x10)
   Variable: M [const; width=8, addr=0]
      Initialization value -- 
      Scalar [[c,p]; width=8] -- width 8: -5 (18446744073709551611,0xfffffffffffffffb)
   Variable: N [const; width=8, addr=0]
      Initialization value -- 
      Scalar [[c,p]; width=8] -- width 8: 300 (300,0x12c)
   Variable: O [const; width=8, addr=0]
      Initialization value -- 
      Scalar [[c,p]; width=8] -- width 8: 14 (14,0xe)
-----------
Functions
-----------
   Function[intrinsic]: str_len() -- 1 args, 0 vars, return type=scalar
   Parameters & Variables -- 
      Variable: ref [param; width=16, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: str_addr() -- 1 args, 0 vars, return type=scalar
   Parameters & Variables -- 
      Variable: ref [param; width=16, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: str_ispkt() -- 1 args, 0 vars, return type=scalar
   Parameters & Variables -- 
      Variable: ref [param; width=16, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: str_seg() -- 1 args, 0 vars, return type=scalar
   Parameters & Variables -- 
      Variable: ref [param; width=16, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: str_isnull() -- 1 args, 0 vars, return type=scalar
   Parameters & Variables -- 
      Variable: ref [param; width=16, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: pkt_new() -- 2 args, 0 vars, return type=void
   Parameters & Variables -- 
      Variable: pnum [param; width=8, addr=0]
      Variable: len [param; width=8, addr=1]
   Body -- 
      (null)
   Function[intrinsic]: pkt_swap() -- 2 args, 0 vars, return type=void
   Parameters & Variables -- 
      Variable: pndst [param; width=8, addr=0]
      Variable: pnsrc [param; width=8, addr=1]
   Body -- 
      (null)
   Function[intrinsic]: pkt_copy() -- 2 args, 0 vars, return type=void
   Parameters & Variables -- 
      Variable: pndst [param; width=8, addr=0]
      Variable: pnsrc [param; width=8, addr=1]
   Body -- 
      (null)
   Function[intrinsic]: pkt_del() -- 1 args, 0 vars, return type=void
   Parameters & Variables -- 
      Variable: pnum [param; width=8, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: pkt_ins_u() -- 3 args, 0 vars, return type=void
   Parameters & Variables -- 
      Variable: pnum [param; width=8, addr=0]
      Variable: off [param; width=8, addr=1]
      Variable: len [param; width=8, addr=2]
   Body -- 
      (null)
   Function[intrinsic]: pkt_ins_d() -- 3 args, 0 vars, return type=void
   Parameters & Variables -- 
      Variable: pnum [param; width=8, addr=0]
      Variable: off [param; width=8, addr=1]
      Variable: len [param; width=8, addr=2]
   Body -- 
      (null)
   Function[intrinsic]: pkt_cut_u() -- 3 args, 0 vars, return type=void
   Parameters & Variables -- 
      Variable: pnum [param; width=8, addr=0]
      Variable: off [param; width=8, addr=1]
      Variable: len [param; width=8, addr=2]
   Body -- 
      (null)
   Function[intrinsic]: pkt_cut_d() -- 3 args, 0 vars, return type=void
   Parameters & Variables -- 
      Variable: pnum [param; width=8, addr=0]
      Variable: off [param; width=8, addr=1]
      Variable: len [param; width=8, addr=2]
   Body -- 
      (null)
   Function[intrinsic]: pkt_parse() -- 1 args, 0 vars, return type=void
   Parameters & Variables -- 
      Variable: pnum [param; width=8, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: parse_push_back() -- 2 args, 0 vars, return type=void
   Parameters & Variables -- 
      Variable: pnum [param; width=8, addr=0]
      Variable: prid [param; width=8, addr=1]
   Body -- 
      (null)
   Function[intrinsic]: parse_pop_back() -- 1 args, 0 vars, return type=void
   Parameters & Variables -- 
      Variable: pnum [param; width=8, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: parse_push_front() -- 2 args, 0 vars, return type=void
   Parameters & Variables -- 
      Variable: pnum [param; width=8, addr=0]
      Variable: prid [param; width=8, addr=1]
   Body -- 
      (null)
   Function[intrinsic]: parse_pop_front() -- 1 args, 0 vars, return type=void
   Parameters & Variables -- 
      Variable: pnum [param; width=8, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: parse_update() -- 1 args, 0 vars, return type=void
   Parameters & Variables -- 
      Variable: pdesc [param; width=8, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: fix_dltype() -- 1 args, 0 vars, return type=void
   Parameters & Variables -- 
      Variable: pnum [param; width=8, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: fix_len() -- 1 args, 0 vars, return type=void
   Parameters & Variables -- 
      Variable: pdesc [param; width=8, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: fix_all_len() -- 1 args, 0 vars, return type=void
   Parameters & Variables -- 
      Variable: pnum [param; width=8, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: fix_csum() -- 1 args, 0 vars, return type=void
   Parameters & Variables -- 
      Variable: pdesc [param; width=8, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: fix_all_csum() -- 1 args, 0 vars, return type=void
   Parameters & Variables -- 
      Variable: pnum [param; width=8, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: meta_get_tstamp() -- 1 args, 0 vars, return type=scalar
   Parameters & Variables -- 
      Variable: pnum [param; width=8, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: meta_set_tstamp() -- 2 args, 0 vars, return type=void
   Parameters & Variables -- 
      Variable: pnum [param; width=8, addr=0]
      Variable: val [param; width=8, addr=1]
   Body -- 
      (null)
   Function[intrinsic]: meta_get_presnap() -- 1 args, 0 vars, return type=scalar
   Parameters & Variables -- 
      Variable: pnum [param; width=8, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: meta_set_presnap() -- 2 args, 0 vars, return type=void
   Parameters & Variables -- 
      Variable: pnum [param; width=8, addr=0]
      Variable: val [param; width=8, addr=1]
   Body -- 
      (null)
   Function[intrinsic]: meta_get_inport() -- 1 args, 0 vars, return type=scalar
   Parameters & Variables -- 
      Variable: pnum [param; width=8, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: meta_set_inport() -- 2 args, 0 vars, return type=void
   Parameters & Variables -- 
      Variable: pnum [param; width=8, addr=0]
      Variable: val [param; width=8, addr=1]
   Body -- 
      (null)
   Function[intrinsic]: meta_get_outport() -- 1 args, 0 vars, return type=scalar
   Parameters & Variables -- 
      Variable: pnum [param; width=8, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: meta_set_outport() -- 2 args, 0 vars, return type=void
   Parameters & Variables -- 
      Variable: pnum [param; width=8, addr=0]
      Variable: val [param; width=8, addr=1]
   Body -- 
      (null)
   Function[intrinsic]: meta_get_flowid() -- 1 args, 0 vars, return type=scalar
   Parameters & Variables -- 
      Variable: pnum [param; width=8, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: meta_set_flowid() -- 2 args, 0 vars, return type=void
   Parameters & Variables -- 
      Variable: pnum [param; width=8, addr=0]
      Variable: val [param; width=8, addr=1]
   Body -- 
      (null)
   Function[intrinsic]: meta_get_class() -- 1 args, 0 vars, return type=scalar
   Parameters & Variables -- 
      Variable: pnum [param; width=8, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: meta_set_class() -- 2 args, 0 vars, return type=void
   Parameters & Variables -- 
      Variable: pnum [param; width=8, addr=0]
      Variable: val [param; width=8, addr=1]
   Body -- 
      (null)
   Function[inline, intrinsic, pconst]: pop() -- 1 args, 0 vars, return type=scalar
   Parameters & Variables -- 
      Variable: num [param; width=8, addr=0]
   Body -- 
      (null)
   Function[inline, intrinsic, pconst]: log2() -- 1 args, 0 vars, return type=scalar
   Parameters & Variables -- 
      Variable: num [param; width=8, addr=0]
   Body -- 
      (null)
   Function[inline, intrinsic, pconst]: min() -- 2 args, 0 vars, return type=scalar
   Parameters & Variables -- 
      Variable: num1 [param; width=8, addr=0]
      Variable: num2 [param; width=8, addr=1]
   Body -- 
      (null)
   Function[inline, intrinsic, pconst]: max() -- 2 args, 0 vars, return type=scalar
   Parameters & Variables -- 
      Variable: num1 [param; width=8, addr=0]
      Variable: num2 [param; width=8, addr=1]
   Body -- 
      (null)
   Function[inline, pconst]: square() -- 1 args, 0 vars, return type=scalar
   Parameters & Variables -- 
      Variable: x [param; width=8, addr=0]
   Body -- 
      Binary Operation: 'multiply' [[p]; width=8]
      Left Operand -- 
         Locator: variable(x) [[p]; width=8]
         Variable -- 
            Variable: x [param; width=8, addr=0]
      Right Operand -- 
         Locator: variable(x) [[p]; width=8]
         Variable -- 
            Variable: x [param; width=8, addr=0]
   Function[inline, pconst]: wierd() -- 2 args, 0 vars, return type=scalar
   Parameters & Variables -- 
      Variable: i [param; width=8, addr=0]
      Variable: j [param; width=8, addr=1]
   Body -- 
      Binary Operation: 'add' [[p]; width=8]
      Left Operand -- 
         Locator: variable(i) [[p]; width=8]
         Variable -- 
            Variable: i [param; width=8, addr=0]
      Right Operand -- 
         Unary Operation: 'negative' [[p]; width=8]
         Operand -- 
            Locator: variable(j) [[p]; width=8]
            Variable -- 
               Variable: j [param; width=8, addr=1]
-----------
Begin Rule
-----------
   (null)
-----------
Packet Rules
-----------
-----------
End Rule
-----------
   (null)
-----------


########
Clearing tree:
########
